File: Leetcode/maximum-subarray.py
class Solution:
  def maxSubArray(self, nums: List[int]) -> int:
    # dp solution time complexity O(n)
    best = Max = nums[0]

    for i in range(1, len(nums)):
      best = max(nums[i], nums[i] + best)
      if best > Max:
        Max = best

    return Max

File: Leetcode/maximum-subarray.cs
public class Solution {
  public int MaxSubArray(int[] nums) {
    int maxSoFar = int.MinValue;
    int maxEndingHere = 0;

    for (int i = 0; i < nums.Length; i++) {
      maxEndingHere += nums[i];

      if (maxSoFar < maxEndingHere) {
        maxSoFar = maxEndingHere;
      }

      if (maxEndingHere < 0) {
        maxEndingHere = 0;
      }
    }
    return maxSoFar;
  }
}